### ISSUE TITLE
Design: out-of-tree platform plugins (vllm-rbln as the first target)

### ISSUE THREAD

**Issue by rebel-jinhwan** [rebel-jinhwan/LLMServingSim#1]
## Summary

vLLM supports non-CUDA hardware through out-of-tree (OOT) platform plugins: a package registers itself under the `vllm.platform_plugins` entry-point group and overrides the platform, scheduler, model runner and attention backend. `vllm-rbln` is one such plugin. LLMServingSim models *vLLM*, but every place where a platform plugin would change vLLM's behaviour is hard-coded to the CUDA answer. This issue proposes a matching plugin seam so that an installed vLLM platform plugin can describe itself to the simulator, and the simulator + profiler + bench pick it up with no CLI flag beyond what the perf bundle already carries.

The goal is that `pip install vllm-rbln` (or a sidecar `llmservingsim-rbln`) is all a user needs, and `python -m profiler`, `python -m bench` and `python -m serving` then behave like vLLM does on that platform.

## Why the current design cannot host vllm-rbln

Traced against `vllm-rbln` `v1/core/rbln_scheduler.py`, `v1/worker/rbln_model_runner.py`, `v1/worker/dp_utils.py` and `v1/worker/bucketing/`:

| Seam | Simulator today (CUDA vLLM) | vllm-rbln | Where it is hard-coded |
|------|-----------------------------|-----------|------------------------|
| Measurement primitive | `layerwise_profile()` CUDA events, per-`nn.Module` | no per-layer CUDA tree; kineto `PrivateUse1` activity exists but the compiled graph is one opaque region | `profiler/core/hooks/extension.py::Extension.fire`, `hooks/timings.py` |
| Latency vs shape | continuous, linear in `tokens` / `sequences` / 4D attention axes | step function: decode batch padded to a bucket, prefill padded to `max_num_tokens`; step latency depends on the *padded* shape | `trace_generator._axis_bracket`, every `_lookup_*` |
| Batch composition | vLLM continuous batching, prefill and decode mixed in one step | **never mixed**: a step is all decodes or a lone prefill (guards A-D in `rbln_scheduler.py`) | `serving/core/scheduler.py::schedule` phase A/B |
| TP collectives | left to ASTRA-Sim (`ALLREDUCE` after `o_proj`/`down_proj`) | inside the compiled graph; a TP=N profile already contains them | `trace_generator._emit_sequence` |
| Engine kwargs / env | `ProfileArgs` + `engine_defaults` | `additional_config`, `VLLM_RBLN_*` env (bucket strategy, buckets) | `profiler/core/config.py`, `bench/core/runner.py` |
| Device identity | `torch.cuda.get_device_name`, `torch.version.cuda` | `rbln-smi` / torch-rbln | `profiler/core/writer.py::_gpu_name`, `_cuda_version` |

Each row is exactly one of the things an OOT plugin overrides in vLLM. That is the argument for mirroring the mechanism rather than adding per-platform `if hardware == "RBLN"` branches.

### 1. Discovery: an entry-point group, and the perf bundle remembers the platform

```toml
# in the plugin's pyproject.toml (vllm-rbln itself, or a sidecar package)
[project.entry-points."llmservingsim.platforms"]
rbln = "vllm_rbln.sim:platform"
```

Resolution order, first hit wins:

1. `--platform <name>` on any of the three CLIs.
2. `meta.yaml::platform` in the perf bundle being loaded (written by the profiler). This is the normal path for `python -m serving`: the bundle says which platform produced it, so a cluster config that names `hardware: RBLN-CA25` needs nothing else.
3. Exactly one installed `llmservingsim.platforms` entry point that is not `cuda` (mirrors vLLM's own `load_plugins_by_group` rule of "one OOT platform plugin at a time").
4. Built-in `cuda`.

The simulator also warns when the runtime configuration disagrees with the profile (bucket list, `max_num_batched_tokens`, `max_num_seqs`) the same way it already does for `engine_effective`.

### 2. Interface: one `Protocol`, every method has the CUDA default

Duck-typed so the plugin package never has to import LLMServingSim.

```python
class SimPlatform(Protocol):
    name: str

    # --- profiler ---------------------------------------------------------
    granularity: Literal["layer", "step"]        # cuda: "layer"
    def measure(self, model_runner, shot, slice_, iterations) -> list[TimingSample]: ...
        # cuda: Extension.fire body (layerwise_profile)
        # rbln: warm-up, then wall-clock execute_model with a device sync,
        #       return one sample named "step"
    def engine_kwargs(self, args: ProfileArgs) -> dict: ...    # cuda: {}
    def device_info(self) -> dict: ...          # name / driver / runtime for meta.yaml
    def sweep_shapes(self, args) -> SweepSpec: ...
        # cuda: current geometric grids
        # rbln: decode axis = every bucket b and b+1; prefill_chunk axis = {max_num_tokens}

    # --- simulator ----------------------------------------------------------
    scheduler_cls: type[Scheduler]              # cuda: Scheduler
        # rbln: subclass with guards A-D (no mixed batching) and
        #       decode batch quantised to the bucket for memory/timing
    comm_in_graph: bool                         # cuda: False -> emit ALLREDUCE/ALLTOALL
                                                # rbln: True  -> profile at real TP, no comm rows
    def kv_bytes_per_token(self, config, fp, kv_fp) -> int: ...  # cuda: current memory_model math

    # --- bench ---------------------------------------------------------------
    def bench_engine_kwargs(self, args) -> dict: ...  # same additional_config / env as profiler
```

`scheduler_cls` is a subclass hook on purpose: that is how vllm-rbln itself does it (`RBLNScheduler(Scheduler)`), and the simulator's `Scheduler` already follows vLLM's `schedule()` shape, so guards A-D port line for line.

### 3. `granularity: "step"` in the trace

With `"step"` the writer emits one extra category `tp<N>/step.csv` keyed like `attention.csv` (`prefill_chunk, kv_prefill, n_decode, kv_decode, time_us`) and no per-layer CSVs. `trace_generator` emits a single `step_0` COMP row per iteration (embedding input `REMOTE`, sampler output `REMOTE` as today) instead of walking `sequence:`. PP is not supported at step granularity in v1 (no block boundary to cut on); the loader raises with a clear message.

Sampling every bucket `b` and `b+1` makes the existing linear `_axis_bracket` reproduce the step function exactly, so no lookup code changes.

### 4. Where the plugin code lives

Recommend **inside vllm-rbln** (`vllm_rbln/sim/`), not a sidecar: the padding and bucketing rules the simulator needs are the same functions the runner uses (`determine_batch_execution_and_padding`, `RBLNBucketingManager`), and keeping them in one repo stops them drifting. The `Protocol` is duck-typed, so vllm-rbln takes no dependency on LLMServingSim. A sidecar remains possible for platforms whose plugin repo we do not control.

## Phasing

Each phase ends with a validation gate against `python -m bench` on real hardware; none of the later phases is worth starting before the earlier gate holds.

- **P0, no simulator code change.** Hand-built RBLN bundle at step granularity by abusing `attention.csv` (whole step time / `num_hidden_layers` per layer, other CSVs zero), plus a `--no-mixed-batching` flag on the current scheduler. Purpose: find out how far the existing lookups and scheduler are from a TP=1 bench run before investing in the seam.
- **P1.** Entry-point discovery, `SimPlatform` protocol, built-in `cuda` platform extracted from the current code paths. `./serving/validate.sh` must show zero moved clocks.
- **P2.** `step.csv` category + `step_0` trace row + `comm_in_graph`. `rbln` platform implemented in vllm-rbln.
- **P3.** Validation ladder on the 8-NPU box: TP=1 single instance, then TP=2/4/8 (profiled at real TP), then DP=8 x TP=1 multi-instance. Calibrate `mem_util` to `kv_cache.num_gpu_blocks` first, per the existing rule. Only after this does predicting configurations the box cannot run (16/32 NPUs, PP) mean anything, and that is when an ASTRA-Sim comm model for RBLN links would be calibrated from measured allreduce times.

## Non-goals

- Modelling the rebel compiler or per-op latency on the NPU; the compiled graph is measured as a whole.
- ASTRA-Sim communication modelling for TP degrees that fit on one box; those are measured.
- Supporting more than one OOT platform in one process (same rule as vLLM).

## Open questions

- Per-step host overhead: RBLN's Python scheduling + input staging is a larger fraction of a step than on GPU, and the simulator has no host-side term. The bench timeseries gives step wall time minus forward time directly; the likely answer is a per-platform constant, but it should be measured before being added.
- Hardware folder name under `profiler/perf/`: derive from `device_info()["name"]` (e.g. `RBLN-CA25`) or keep it a free CLI label as today.
- vLLM version drift: the simulator scheduler is a port of vLLM 0.19's `schedule()`, the bench now runs 0.24. Worth a diff of `vllm/v1/core/sched/scheduler.py` between the two before P3 so a scheduler difference is not misread as a platform difference.
- Should `meta.yaml::platform` be a hard requirement for a bundle to load, or a soft default to `cuda` for existing bundles? Soft default is the backward-compatible choice.

**Comment by rebel-jinhwan** [comment]
Implemented in #2, with two deviations from this design:

1. **No ported scheduler.** Instead of subclassing the simulator's scheduler with vllm-rbln's guards, `serving/core/vllm_scheduler.py` drives vLLM's own scheduler classes the way `EngineCore` does on the host, and the `rbln` platform pins `scheduler_cls` to `vllm_rbln.v1.core.rbln_scheduler.RBLNScheduler`. Against upstream vLLM 0.24 on CPU the in-tree port produces identical batches (856 steps, prefix caching on and off); under KV pressure the only difference is vLLM's reserved null block.
2. **Bucket rounding instead of `b`/`b+1` sampling.** `_lookup_step` rounds a decode batch up to the next profiled `n_decode` and snaps a prefill chunk to the profiled chunk, reading the shapes off `step.csv`.

First result on 4x RBLN-CR03, MiniMax-M2.5 tp4+EP, 24 requests: raw bundle TTFT/TPOT mean -7.3% / -3.9%; with the two host-overhead knobs (`--step-overhead-us 1100 --prefill-step-overhead-us 9000`) +0.0% / +0.7%. That confirms the open question above: the missing piece was per-step host time, and it is larger on a prefill step than on a decode step (about 10 ms vs 1 ms).

Layout ended up as `platforms/<vendor>/{__init__,profile,simulator}.py` rather than a single Protocol object, so the profiler side never imports in the simulator container and vice versa.
